forte/data/data_pack_dataset.py

```diff
@@ -0,0 +1,278 @@
+                 skip_k: int = 0):
+        self._get_data_args: Dict = {
+            "context_type": context_type,
+            "request": request,
+            "skip_k": skip_k
+        }
+
+        self._data_pack_iter: Iterator[DataPack] = pack_generator
+        self._instance_iter: Optional[Iterator[Instance]] = None
+        self._curr_data_pack: Optional[DataPack] = None
+
+    def __iter__(self):
+        return self
+
+    def __next__(self) -> RawExample:
+        if self._curr_data_pack is None:
+            self._curr_data_pack = next(self._data_pack_iter)
+            self._instance_iter = \
+                self._curr_data_pack.get_data(**self._get_data_args)
+
+        assert self._instance_iter is not None
+
+        try:
+            return next(self._instance_iter), self._curr_data_pack
+        except StopIteration:
+            # Current data pack has no more instance. Go to next data pack.
+            self._curr_data_pack = next(self._data_pack_iter)
+            self._instance_iter = \
+                self._curr_data_pack.get_data(**self._get_data_args)
+
+        return next(self._instance_iter), self._curr_data_pack
+
+
+class DataPackDataSource(IterDataSource):
+    """
+    A data source consists of data packs. It contains an iterator over
+    :class:`forte.data.data_pack_dataset.DataPackIterator`.
+
+    Args:
+        pack_generator (Iterator[DataPack]): A generator of
+            :class:`forte.data.data_pack.DataPack`.
+        context_type: The granularity of a single example which
+            could be any ``Annotation`` type. For example, it can be
+            :class:`ft.onto.base_ontology.Sentence`, then each training example
+            will represent the information of a sentence.
+        request: The request of type `Dict` sent to
+            :class:`forte.data.readers.base_reader.PackReader` to query
+            specific data.
+        skip_k (int): Will skip the first `skip_k` instances and generate
+            data from the (`offset` + 1)th instance.
+
+    .. note::
+        For parameters `context_type`, `request`, `skip_k`, please refer to
+        :meth:`get_data()` in :class:`forte.data.data_pack.DataPack`.
+    """
+    def __init__(self,
+                 pack_generator: Iterator[DataPack],
+                 context_type: Type[Annotation],
+                 request: Optional[DataRequest] = None,
+                 skip_k: int = 0):
+        self._iterator: Iterator = DataPackIterator(pack_generator,
+                                                    context_type,
+                                                    request,
+                                                    skip_k)
+        super().__init__(self)
+
+    def __iter__(self):
+        return self._iterator
+
+
+class DataPackDataset(DatasetBase):
+    """
+    A dataset representing data packs. Calling an
+    `DataIterator
+    <https://texar-pytorch.readthedocs.io/en/latest/code/data.html#dataiterator>`_
```

Comment:
We can do inter-sphinx link here with texar-pytorch

Reply:
Fixed.